### DISCUSSION TITLE
How to run an Appium test case in sequence on two real Android devices?

### DISCUSSION THREAD

**Discussion by sivapriyav** [appium/appium#20629] [General]
I'm trying to automate a messaging scenario for an android application using Appium where Device1 sends a message to Device2. My goal is to run the test on two real Android devices. Here's the flow I want to implement:

1. Start the Appium server on ports 4723 and 4725, connecting the two devices to different ports
2. Initialize the device IDs and other desired capabilities.
3. Initialize two Appium drivers on ports 4723 and 4725.
4. Send a message from Device1 using driver1 and verify the receipt of the message in Device2 using driver2.

**Issue**
The test case only executes on one device instead of both.

All execution is performed on real devices.

Question: What do I need to make in order to run this test case in both devices? Any guidance on using threading or any other approach would be greatly appreciated!

Adding a sample code

```
from appium import webdriver
from appium.options.android import UiAutomator2Options
from appium.webdriver.common.appiumby import AppiumBy
import time

def send_message_on_device_1(driver):
    """ Function to send a message from device 1 """
    # Code to send message

def verify_message_on_device_2(driver):
    """ Function to verify the message on device 2 """
    # Code to verify message

def initialize_driver(device_name, port):
    """ Function to initialize Appium driver for each device """
    desired_caps = {
        "platformName": "Android",
        "deviceName": device_name,
        "appPackage": "your.app.package",
        "appActivity": "your.app.activity",
        "automationName": "UiAutomator2",
        "noReset": True
    }
    capabilities_options = UiAutomator2Options().load_capabilities(desired_caps)
    appium_server_url = f'http://127.0.0.1:{port}/wd/hub'
    driver = webdriver.Remote(command_executor=appium_server_url, options=capabilities_options)
    return driver

if __name__ == "__main__":
    # Initialize Appium driver for device 1 on port 4723
    driver1 = initialize_driver("device1_udid", 4723)

    # Initialize Appium driver for device 2 on port 4725
    driver2 = initialize_driver("device2_udid", 4725)

    try:
        # Send message from device 1
        send_message_on_device_1(driver1)

        # Verify message on device 2
        verify_message_on_device_2(driver2)

    finally:
        # Quit both drivers
        driver1.quit()
        driver2.quit()
```

**Comment by KazuCocoa** [comment]
Did you take a look at https://github.com/appium/appium-uiautomator2-driver?tab=readme-ov-file#parallel-tests? Asking such a question in https://discuss.appium.io/ would also help.

**Reply by sivapriyav** [reply]
> Did you take a look at https://github.com/appium/appium-uiautomator2-driver?tab=readme-ov-file#parallel-tests? Asking such a question in https://discuss.appium.io/ would also help.
Thank you! I hadn’t come across this doc before, but adding the systemport capability worked perfectly for me. I’ll also consider posting on the Appium forum for further insights. Thanks again for the suggestion!
